infer/environment.cc

```diff
@@ -324,24 +436,34 @@ void Environment::propagateKnowledge(core::Context ctx, cfg::LocalRef to, cfg::L
             toState.typeAndOrigins.type = core::Types::nilClass();
         }
 
-        toKnowledge = fromKnowledge;
-        toKnowledge.truthy.mutate().noTypeTests.emplace_back(from, core::Types::falsyTypes());
-        toKnowledge.falsy.mutate().yesTypeTests.emplace_back(from, core::Types::falsyTypes());
-        fromKnowledge.truthy.mutate().noTypeTests.emplace_back(to, core::Types::falsyTypes());
-        fromKnowledge.falsy.mutate().yesTypeTests.emplace_back(to, core::Types::falsyTypes());
+        // Copy properties from fromKnowledge to toKnowledge
+        toKnowledge.replace(to, typeTestsWithVar, fromKnowledge);
+
+        toKnowledge.truthy().addNoTypeTest(to, typeTestsWithVar, from, core::Types::falsyTypes());
+        toKnowledge.falsy().addYesTypeTest(to, typeTestsWithVar, from, core::Types::falsyTypes());
+        fromKnowledge.truthy().addNoTypeTest(from, typeTestsWithVar, to, core::Types::falsyTypes());
+        fromKnowledge.falsy().addYesTypeTest(from, typeTestsWithVar, to, core::Types::falsyTypes());
         fromKnowledge.sanityCheck();
         toKnowledge.sanityCheck();
     }
 }
 
 void Environment::clearKnowledge(core::Context ctx, cfg::LocalRef reassigned, KnowledgeFilter &knowledgeFilter) {
-    for (auto &el : vars) {
-        auto &k = el.second.knowledge;
-        if (knowledgeFilter.isNeeded(el.first)) {
-            k.removeReferencesToVar(reassigned);
-            k.sanityCheck();
+    auto &typesWithReassigned = typeTestsWithVar.get(reassigned);
+    if (!typesWithReassigned.empty()) {
```

Comment:
Just checking my understanding here, this conditional is the optimization introduced?

Reply:
This conditional _and_ the inner loop _only_ checks local refs that at one point referenced the reassigned localref (rather than all vars).